refactor: use standard logger instead of logrus

lzap opened the pull request:
I would like to propose moving away from logrus in the image-builder-cli (and standalone CLI tools from composer and images as well). The library is in maintenance mode for quite a while and I believe it does not provide a good structured logging capabilities and API (log/slog is better) while for console logging it is an overkill. It was primarily designed to be structured logging library with some human-formatting features (Printf...).

Now, for services like crc/composer, I am currently working on replacing logrus with log/slog structured logging library which is part of Go. However, structured logging is not a great fit for CLI applications since its primary purpose is to machine-readable output (JSON) and although text formatters are available in the standard library, it is not best reading for humans. Example: time=2025-04-29 11:30 msg=Initializing osbuild

The question tho is what to replace logrus with, and what I would like to propose is a simple one: there is a log package in the standard library that is extremely simple. It basically only provides Print / Prinln / Printf plus Error/Panic counterparts which do the same but exits the program. There are no logging levels, because they do not make sense in the console UNIX apps context - programs should be silent by default and diagnostic output or errors should be printed to stderr. The standard logger in Go is configured to send all output to stderr actually. I like this simplicity. Also, the API is so clean and small that it is super easy to create a utility package that can provide multiple loggers with multiple "levels" and outputs if needed.

The initial version of the PR comes with a new package called `olog` (optional log) which stands for optional logger and it is a proxy to `log.Logger` that can be turned off or on depending on the `--verbose` option. While this could have been a simple variable somewhere in the code, its own package enables seamless and short use: `olog.Printf(...)` that nicely match `log.Printf(...)`.

My plan is to the similar PR in the `images` library with similar `olog` package (a copy is perhaps sufficient this API is stable). This way, the logging can be enabled, disabled or redirected from API users.

mvo5 commented on pkg/setup/setup.go:
```diff
@@ -136,7 +135,7 @@ func validateCanRunTargetArch(targetArch string) error {
 		// any arch should work so let's just warn. the common case
 		// (arm64/amd64) is covered properly
-		logrus.Warningf("cannot check architecture support for %v: no canary binary found", targetArch)
+		log.Panicf("cannot check architecture support for %v: no canary binary found", targetArch)
```
Please do not change a warning into a panic in the middle of a refactor, if we want to do this, a change like this should be its own commit with a proper rational.

lzap replied:
Accidentally changed behavior from just a warning to panic, will fix.

mvo5 commented on cmd/image-builder/main.go:
```diff
@@ -409,13 +411,11 @@ func cmdDescribeImg(cmd *cobra.Command, args []string) error {
-	// messages.
-	logrus.SetOutput(io.Discard)
+	// Forward to the optional logger that can be turned on via options.
+	logrus.SetOutput(olog.Default().Writer())
```
Do we still need this? Could we just remove logrus entirely?

lzap replied:
I will go ahead today and create a PR in images too, then depending on the order of merges I will update this. Or if it is okay to see some noisy logrus output for a while, I can remove this straight away.

lzap replied:
Since I already filed a PR for images, I am removing this line and `logrus` completely from this project, adding this commented-out line (can be uncommented once the images PR is merged):

```go
	if verbose {
		olog.SetDefault(log.New(os.Stderr, "", 0))
		// XXX: add once images has olog support
		//images_log.SetDefault(log.New(os.Stderr, "", 0))
	}
```

achilleas-k replied:
Great!

mvo5 replied:
Aha, thanks! I missed that without it we get all the verbosity from "images". I guess we can just make sure images gets merged first and then we update.

lzap commented on pkg/util/util.go:
```diff
@@ -17,7 +17,7 @@ func IsMountpoint(path string) bool {
 // to the current process's stdout and stderr
 func RunCmdSync(cmdName string, args ...string) error {
-	logrus.Debugf("Running: %s %s", cmdName, strings.Join(args, " "))
+	olog.Printf("Running: %s %s", cmdName, strings.Join(args, " "))
```
This is optional log statement, it will only appear when `--verbose` is provided.

mvo5 commented on cmd/image-builder/main.go:
```diff
@@ -520,12 +520,19 @@ operating systems like Fedora, CentOS and RHEL with easy customizations support.
 	if err := run(); err != nil {
-		fmt.Fprintf(osStderr, "error: %s\n", err)
-		os.Exit(1)
+		log.Fatalf("error: %s\n", err)
```
When quickly playing with this it seems by default we now output with data/time prefix:
```console
$ go build && ./image-builder xxx
2025/04/29 17:15:01 error: unknown command "xxx" for "image-builder"
```
I would prefer if we would not do this by default as it adds quite a bit of visual noise.

lzap replied:
Sure, this is configurable. Let me do that.

lzap replied:
Fixed:

```
lzap@dev:~/image-builder-cli$ go run ./cmd/image-builder/ xxx 2>/dev/null

lzap@dev:~/image-builder-cli$ go run ./cmd/image-builder/ xxx
error: unknown command "xxx" for "image-builder"
exit status 1
```

mvo5 commented on cmd/image-builder/main.go:
```diff
@@ -520,12 +520,19 @@ operating systems like Fedora, CentOS and RHEL with easy customizations support.
 
 	rootCmd.AddCommand(describeImgCmd)
 
+	verbose, err := rootCmd.Flags().GetBool("verbose")
```
This will need to be  `Flags`->`PersientFlags`

mvo5 commented on cmd/image-builder/main.go:
```diff
@@ -520,12 +520,19 @@ operating systems like Fedora, CentOS and RHEL with easy customizations support.
 
+	verbose, err := rootCmd.Flags().GetBool("verbose")
+	if err != nil {
+		log.Panicln(err)
```
I personally would just return the err here but of course that is a place where a panic is also apporpriate

mvo5 commented on internal/log/olog/proxy.go:
```diff
@@ -0,0 +1,83 @@
+
+import "io"
+
+// SetOutput is a proxy for log.SetOutput.
```
It seems the only parts we use from this currently is "Printf()", can we YAGNI this and just remove everything here (with a comment maybe) and just keep "Prinf()" for now and add as needed?

lzap replied:
Yeah I wanted to implement all methods because I very often see APIs defining interface for the standard library logger struct in various detail, some only have `Print` functions, others include `Panic` and some actually have all of the methods.

This API will not change anytime soon, I am pretty sure. So we could just leave that there. Or a compromise: just keep all the logging methods and leave setters/getters, except `Writer` that is a useful one.

mvo5 reviewed (commented):
Thanks, very excited that we can get rid of logrus! A few inline questions/comments, maybe worthwhile to have a high-bandwidth sync.

achilleas-k commented on internal/log/package.go:
```diff
@@ -0,0 +1,2 @@
+// Logging utility functions
+package log
```
Is the nesting necessary?  Couldn't we have `internal/olog` be the package, instead of `internal/log/olog`?

lzap replied:
Sure thing.

lzap commented:
So I rebased and moved from `log/olog` to `olog`. The default logger does not have any time/date prefixes now (that was an overlook I hate this too). I cleaned up the new package and broken it down into two files - one is the variable and the other is for the proxy type.

Fixed all the bugs or typos, the initial version should have been draft PR but I was in a rush, this is now much better.

lzap commented:
There was an alternative proposal to simply just use `log/slog` package and configure it to print just message. A custom handler would be needed for that, but the problem is that this library is strictly for structured logging and the issue is the output would have to be structured. If we only include the `msg` field, then we are loosing a lot, example:

    slog.Info("user created", "username", name)

Would only print

    user created

leaving the additional (and very important) data behind. The only option I see is to format it something like `msg="user created" user="lzap"` or in similar fashion.

Just wanted to drop it here, I think using the simple `log` built-in package is a reasonable approach.

mvo5 commented on pkg/progress/progress.go:
```diff
@@ -241,7 +241,7 @@ func (b *terminalProgressBar) Stop() {
 		// closing would not work but lets be conservative -
 		// without a timeout we hang here forever
-		logrus.Warnf("no progress channel shutdown after 1sec")
+		log.Printf("no progress channel shutdown after 1sec")
```
Please add a `WARNING:` prefix here for the message

mvo5 commented on pkg/progress/progress.go:
```diff
@@ -479,7 +479,7 @@ func runOSBuildWithProgress(pb ProgressBar, manifest []byte, exports []string, o
 		for p := st.Progress; p != nil; p = p.SubProgress {
 			if err := pb.SetProgress(i, p.Message, p.Done, p.Total); err != nil {
-				logrus.Warnf("cannot set progress: %v", err)
+				log.Printf("cannot set progress: %v", err)
```
Same here, please just add a `WARNING:` prefix to the string

mvo5 commented on pkg/setup/setup.go:
```diff
@@ -136,7 +135,7 @@ func validateCanRunTargetArch(targetArch string) error {
 		// any arch should work so let's just warn. the common case
 		// (arm64/amd64) is covered properly
-		logrus.Warningf("cannot check architecture support for %v: no canary binary found", targetArch)
+		log.Printf("cannot check architecture support for %v: no canary binary found", targetArch)
```
And here :)

mvo5 commented on pkg/setup/setup_test.go:
```diff
@@ -22,12 +22,14 @@ func TestValidateCanRunTargetArchTrivial(t *testing.T) {
 func TestValidateCanRunTargetArchUnsupportedCanary(t *testing.T) {
-	var logbuf bytes.Buffer
-	logrus.SetOutput(&logbuf)
+	logbuf := new(bytes.Buffer)
```
I'm curious (and its not important), any particular reason to go from `var logbuf bytes.Buffer` to `logbuf := new(bytes.Buffer)` ? Is there a technical (or other) advantage or just (personal) preference? No need to change it, I don't mind - just curious if there is something to learn here :) 

I mean, from a purely technical PoV having `var logbuf` on the stack seems "cheaper" because no malloc() call is require and the GC does not need to track it (as its just on the stack and auto cleaned).

lzap replied:
Good catch, no, no reason. Escape analysis is a strong feature we should always take advantage of. Removed here and there (total 2 places).

mvo5 commented on cmd/image-builder/main.go:
```diff
@@ -520,12 +516,21 @@ operating systems like Fedora, CentOS and RHEL with easy customizations support.
+	if verbose {
+		olog.SetDefault(log.New(os.Stderr, "", 0))
+		// XXX: add once images has olog support
+		//images_log.SetDefault(log.New(os.Stderr, "", 0))
```
We can enable this now (as the  images PR is now merged), right? But I guess this can also be a followup

lzap replied:
Oh great eyes, yes, let me rebase.

lzap replied:
Oh wait we need to upgrade images, do we? Probably a followup, yeah. Let me create a ticket:

https://github.com/osbuild/image-builder-cli/issues/216

mvo5 reviewed (approved):
Thank you
